### PR TITLE
Started handling messages from background isolates for Android

### PR DESCRIPTION
**Pull request by gaaclarke**
This provides all the work for handling platform channels in the Dart->Host direction for Android.  For more details and integration tests see the framework PR.

Integration tests exist in the framework PR.

companion framework PR: https://github.com/flutter/flutter/pull/111279
companion engine PR (iOS): https://github.com/flutter/engine/pull/35174
issue: https://github.com/flutter/flutter/issues/13937

## Pre-launch Checklist

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [x] I read and followed the [Flutter Style Guide] and the [C++, Objective-C, Java style guides].
- [x] I listed at least one issue that this PR fixes in the description above.
- [x] I added new tests to check the change I am making or feature I am adding, or Hixie said the PR is test-exempt. See [testing the engine] for instructions on
writing and running engine tests.
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] I signed the [CLA].
- [x] All existing and new tests are passing.

If you need help, consider asking for advice on the #hackers-new channel on [Discord].

[Contributor Guide]: https://github.com/flutter/flutter/wiki/Tree-hygiene#overview
[Tree Hygiene]: https://github.com/flutter/flutter/wiki/Tree-hygiene
[Flutter Style Guide]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo
[C++, Objective-C, Java style guides]: https://github.com/flutter/engine/blob/main/CONTRIBUTING.md#style
[testing the engine]: https://github.com/flutter/flutter/wiki/Testing-the-engine
[CLA]: https://cla.developers.google.com/
[flutter/tests]: https://github.com/flutter/tests
[breaking change policy]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes
[Discord]: https://github.com/flutter/flutter/wiki/Chat

### CODE REVIEW

**Comment by gaaclarke** [comment]
@dnfield there isn't much code that is changed here, mostly comments and turning one integer atomic.  If you want me to walk through why this is safe let me know.  I've added comments to where the threading policy is changing so you can read through to make sure it's safe.  My testing plan is to rely on the integration tests in the framework PR.

**Comment by gaaclarke** [comment]
> It looks like this pull request may not have tests. Please make sure to add tests before merging. If you need [an exemption](https://github.com/flutter/flutter/wiki/Tree-hygiene#tests) to this rule, contact Hixie on the #hackers channel in [Chat](https://github.com/flutter/flutter/wiki/Chat) (don't just cc him here, he won't see it! _He's on Discord!_).

Integration tests are in the framework PR.

File: shell/platform/android/io/flutter/embedding/engine/dart/DartMessenger.java [hunk 1]
```diff
@@ -338,11 +339,13 @@ private void dispatchMessageToQueue(
...
   public void handleMessageFromDart(
       @NonNull String channel, @Nullable ByteBuffer message, int replyId, long messageData) {
-    // Called from the ui thread.
+    // Called from any thread.
     Log.v(TAG, "Received message from Dart over channel '" + channel + "'");
 
     HandlerInfo handlerInfo;
     boolean messageDeferred;
+    // TODO(gaaclarke): This can potentially be a bottleneck and could replaced
+    // with a read/write lock.
```
Comment (dnfield) [inline shell/platform/android/io/flutter/embedding/engine/dart/DartMessenger.java:348]: file a bug

**Review by dnfield — APPROVED** [review]

File: shell/platform/android/io/flutter/embedding/engine/dart/DartMessenger.java (hunk 1 above)
Comment (gaaclarke) [reply shell/platform/android/io/flutter/embedding/engine/dart/DartMessenger.java]: I switched it to a regular comment if you don't mind.  I'm not sure I want to commit to the idea that it is something we need to do, but I don't want to lose the idea that it was something we could do.

Comment (dnfield) [reply shell/platform/android/io/flutter/embedding/engine/dart/DartMessenger.java]: I think it'd be fine to file a bug either way to add a benchmark for this and figure out if/how much a different lock would help.  But non-blocking for landing this.

Comment (gaaclarke) [reply shell/platform/android/io/flutter/embedding/engine/dart/DartMessenger.java]: https://github.com/flutter/flutter/issues/111489

**Pull request merged**